vshard/replicaset.lua

```diff
@@ -144,6 +144,49 @@ local function netbox_wait_connected(conn, timeout)
     return timeout
 end
 
+local netbox_is_conn_dead
+
+if not (util.version_is_at_least(2, 10, 0, 'beta', 2, 0) and
+        util.version_is_at_most(2, 10, 0, nil, 0, 0)) then
+
+--
+-- Check if the connection is dead and it won't be restored automatically.
+-- Even though replicaset's connections are initialized with the option
+-- 'reconnect_after', there's cases, when no reconnect will be done (e.g.
+-- if the conn was explicitly cancelled or the connection's worker fiber
+-- was killed). In these situations we need to reestablish it manually as
+-- the missing connection to some replicaset is a critical problem which
+-- leads to the inability of the user to access some part of a data via
+-- router.
+--
+-- Note: the function expects the conn to be non-nil.
+--
+netbox_is_conn_dead = function(conn)
+    return conn.state == 'error' or conn.state == 'closed'
+end
+
+else -- 2.10.0
+
+--
+-- Tarantool 2.10.0 has a bug: after worker fiber kill connection's state is
+-- error_reconnect, but it doesn't reconnet anymore. The only way we can
+-- catch this situation is error string comparison as the connection doesn't
+-- have _fiber in itself.
+--
+netbox_is_conn_dead = function(conn)
+    if conn.state == 'error' or conn.state == 'closed' then
```

Comment:
I would suggest to move `if conn.state == 'active' then return true ... ` right here. Otherwise you force the function's user to explicitly check the connection's state before calling `netbox_is_conn_dead()` if he wants a quick path for the common case.

I know I proposed myself to check for `'active'` state explicitly outside, but now when I look at it, I think better make `netbox_is_conn_dead()` have the happy path as well. Up to you.